src/coprocessor/dag/expr/builtin_compare.rs

```diff
@@ -158,6 +160,114 @@ impl ScalarFunc {
         do_coalesce(self, |v| v.eval_json(ctx, row))
     }
 
+    pub fn greatest_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        do_get_extremum(self, |e| e.eval_int(ctx, row), max)
+    }
+
+    pub fn greatest_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        do_get_extremum(self, |e| e.eval_real(ctx, row), |x, y| x.max(y))
+    }
+
+    pub fn greatest_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), max)
+    }
+
+    pub fn greatest_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut greatest = mysql::time::zero_datetime(ctx.cfg.tz);
+
+        for exp in &self.children {
+            let s = try_opt!(exp.eval_string_and_decode(ctx, row));
+            match Time::parse_datetime(&s, Time::parse_fsp(&s), ctx.cfg.tz) {
+                Ok(t) => greatest = max(greatest, t),
+                Err(_) => {
+                    if let Err(e) = ctx.handle_invalid_time_error(Error::invalid_time_format(&s)) {
+                        return Err(e);
+                    }
+                }
+            }
+        }
+
+        Ok(Some(Cow::Owned(greatest.to_string().into_bytes())))
+    }
+
+    pub fn greatest_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        do_get_extremum(self, |e| e.eval_string(ctx, row), max)
+    }
+
+    pub fn least_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        do_get_extremum(self, |e| e.eval_int(ctx, row), min)
+    }
+
+    pub fn least_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        do_get_extremum(self, |e| e.eval_real(ctx, row), |x, y| x.min(y))
+    }
+
+    pub fn least_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), min)
+    }
+
+    pub fn least_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut res = Cow::from(vec![]);
+        let mut least = Time::new(
+            ctx.cfg.tz.timestamp(
+                mysql::time::MAX_TIMESTAMP,
+                mysql::time::MAX_TIME_NANOSECONDS,
+            ),
+            mysql::types::DATETIME,
+            mysql::MAX_FSP,
+        )?;
+        let mut find_invalid_time = false;
+
+        for exp in &self.children {
+            let val = try_opt!(exp.eval_string(ctx, row));
+            let s = super::decode(val.clone(), exp.get_tp().get_charset())?;
+            // let s = try_opt!(exp.eval_string_and_decode(ctx, row));
```

Comment:
Why not use `eval_string_and_decode` instead?

Reply:
because I want to keep `val` here

Reply:
Could we use `s.into_bytes` instead?